nvflare/apis/impl/controller.py

```diff
@@ -11,474 +11,43 @@
-
-    if targets is not None:
-        if not isinstance(targets, list):
-            raise TypeError("targets must be a list of Client or string, but got {}".format(type(targets)))
-
-        for t in targets:
-            if not isinstance(t, (Client, str)):
-                raise TypeError(
-                    "targets must be a list of Client or string, but got element of type {}".format(type(t))
-                )
-
-
-def _get_client_task(target, task: Task):
-    for ct in task.client_tasks:
-        if target == ct.client.name:
-            return ct
-    return None
-
-
-class Controller(Responder, ControllerSpec, ABC):
+class Controller(FLComponent, ControllerSpec, ABC):
     def __init__(self, task_check_period=0.2):
-        """Manage life cycles of tasks and their destinations.
+        """Controller logic for tasks and their destinations.
+
+        Must set_communicator() to access communication related function implementations.
 
         Args:
-            task_check_period (float, optional): interval for checking status of tasks. Defaults to 0.2.
+            task_check_period (float, optional): interval for checking status of tasks. Applicable for WFCommServer. Defaults to 0.2.
         """
         super().__init__()
-        self._engine = None
-        self._tasks = []  # list of standing tasks
-        self._client_task_map = {}  # client_task_id => client_task
-        self._all_done = False
-        self._task_lock = Lock()
-        self._task_monitor = threading.Thread(target=self._monitor_tasks, args=())
         self._task_check_period = task_check_period
-        self._dead_client_reports = {}  # clients that reported the job is dead on it: name => report time
-        self._dead_clients_lock = Lock()  # need lock since dead_clients can be modified from different threads
-        # make sure _check_tasks, process_task_request, process_submission does not interfere with each other
-        self._controller_lock = Lock()
-
-    def initialize_run(self, fl_ctx: FLContext):
-        """Called by runners to initialize controller with information in fl_ctx.
+        self.communicator = None
 
-        .. attention::
+    def set_communicator(self, communicator: WFCommSpec, fl_ctx: FLContext):
+        if not isinstance(communicator, WFCommSpec):
+            raise TypeError(f"communicator must be an instance of WFCommSpec, but got {type(communicator)}")
 
-            Controller subclasses must not overwrite this method.
-
-        Args:
-            fl_ctx (FLContext): FLContext information
-        """
+        self.communicator = communicator
+        self.communicator.controller = self
+        self.communicator.task_check_period = self._task_check_period
         engine = fl_ctx.get_engine()
         if not engine:
             self.system_panic(f"Engine not found. {self.__class__.__name__} exiting.", fl_ctx)
             return
 
         self._engine = engine
```

Comment:
I think we should separate the `set_communicator` logic VS `set_engine/initialize` logic, as I argued below.